### PR TITLE
[SPARK-6592][SQL] fix filter for scaladoc to generate API doc for Row class under catalyst dir

### PR DESCRIPTION
**Pull request by CodingCat**
https://issues.apache.org/jira/browse/SPARK-6592

The current impl in SparkBuild.scala filter all classes under catalyst directory, however, we have a corner case that Row class is a public API under that directory

we need to include Row into the scaladoc while still excluding other classes of catalyst project

Thanks for the help on this patch from @rxin and @liancheng

### CODE REVIEW

File: project/SparkBuild.scala [hunk 1]
```diff
@@ -366,7 +366,7 @@ object Unidoc {
       .map(_.filterNot(_.getCanonicalPath.contains("executor")))
       .map(_.filterNot(_.getCanonicalPath.contains("python")))
       .map(_.filterNot(_.getCanonicalPath.contains("collection")))
-      .map(_.filterNot(_.getCanonicalPath.contains("sql/catalyst")))
+      .map(_.filterNot(_.getCanonicalPath.contains("apache/spark/sql/catalyst")))
       .map(_.filterNot(_.getCanonicalPath.contains("sql/execution")))
```
Comment (rxin) [inline project/SparkBuild.scala]: in light of this, i think we should tighten all the paths. can you put the whole path path for all of the packages listed here?
e.g. org/apache/spark/sql/catalyst, org/apache/spark/executor, ..

Make sure you build it to verify it works as expected.

Thanks.

**Comment by CodingCat** [comment]
@rxin , thanks for the comments

I change the code in SparkBuild and found there are several related problems in the current code 


1. the original SparkBuild filters all classes under `collection ` . However, some of the classes there are marked as `DeveloperAPI`, shouldn't we document all DeveloperAPIs? but if so, I found that some classes are DeveloperAPI but meanwhile they are `private[spark]`, so they are not documented


2. regarding the classes under `org.apache.spark.sql.types`, nearly all classes defined in dataTypes.scala are 'DeveloperAPIs' but they are excluded from the scaladoc originally, similar with 1, we should actually document them, right? (in the current version of patch, I exclude them temporarily)

If we do document classes under `org.apache.spark.sql.types`,  for the other helper classes in the same package, e.g. DataTypeConversions, shall we document them as well? (it is `protected[sql]` now)

also cc: @liancheng and @marmbrus

**Comment by rxin** [comment]
It's ok let's ignore those collections for now. I don't think they are really meant to be public public.

**Comment by rxin** [comment]
sq..types should be documented.

File: project/SparkBuild.scala [hunk 2]
```diff
@@ -360,15 +360,16 @@ object Unidoc {
...
+      .map(_.filterNot(_.getCanonicalPath.contains("org/apache/spark/shuffle")))
+      .map(_.filterNot(_.getCanonicalPath.contains("org/apache/spark/executor")))
       .map(_.filterNot(_.getCanonicalPath.contains("python")))
-      .map(_.filterNot(_.getCanonicalPath.contains("collection")))
-      .map(_.filterNot(_.getCanonicalPath.contains("sql/catalyst")))
-      .map(_.filterNot(_.getCanonicalPath.contains("sql/execution")))
-      .map(_.filterNot(_.getCanonicalPath.contains("sql/hive/test")))
+      .map(_.filterNot(_.getCanonicalPath.contains("org/apache/spark/collection")))
+      .map(_.filterNot(_.getCanonicalPath.contains("org/apache/spark/sql/catalyst")))
+      .map(_.filterNot(_.getCanonicalPath.contains("org/apache/spark/sql/types")))
```
Comment (rxin) [inline project/SparkBuild.scala]: let's not ignore types

**Comment by CodingCat** [comment]
I committed more modifications, generally 

1. ignore collections

2. enable types, **BUT** changed those `protected[sql] ` to `private[sql]` to exclude them from scaladoc

**Comment by sryza** [comment]
Mind tagging this with SQL so it gets sorted properly?

**Comment by CodingCat** [comment]
done

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
Thanks. I'm merging this in master & branch-1.3.

**Comment by CodingCat** [comment]
thanks for the review

**Pull request closed without merging**